File: backend/src/main/java/com/northwollo/tourism/controller/AdminAuditController.java
package com.northwollo.tourism.controller;

import com.northwollo.tourism.entity.AuditLogEntry;
import com.northwollo.tourism.service.AuditLogService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/admin/audit")
@RequiredArgsConstructor
@Tag(name = "Admin Audit Management", description = "Admin endpoints for audit log management")
@SecurityRequirement(name = "bearerAuth")
@PreAuthorize("hasRole('ADMIN')")
public class AdminAuditController {

    private final AuditLogService auditLogService;

    @GetMapping
    @Operation(summary = "Get all audit logs with pagination")
    public ResponseEntity<Page<AuditLogEntry>> getAllAuditLogs(
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        // Use sorting by timestamp descending
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "timestamp"));
        Page<AuditLogEntry> auditLogs = auditLogService.getAuditLogs(pageable);
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/search")
    @Operation(summary = "Search audit logs with multiple criteria")
    public ResponseEntity<Page<AuditLogEntry>> searchAuditLogs(
            @Parameter(description = "User ID filter") @RequestParam(required = false) Long userId,
            @Parameter(description = "Username filter") @RequestParam(required = false) String username,
            @Parameter(description = "Action filter") @RequestParam(required = false) String action,
            @Parameter(description = "Resource type filter") @RequestParam(required = false) String resourceType,
            @Parameter(description = "Category filter") @RequestParam(required = false) String category,
            @Parameter(description = "Severity filter") @RequestParam(required = false) String severity,
            @Parameter(description = "IP address filter") @RequestParam(required = false) String ipAddress,
            @Parameter(description = "Start time filter") @RequestParam(required = false) 
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime startTime,
            @Parameter(description = "End time filter") @RequestParam(required = false) 
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime endTime,
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<AuditLogEntry> auditLogs = auditLogService.searchAuditLogs(
                userId, username, action, resourceType, category, severity, ipAddress,
                startTime, endTime, pageable);
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/user/{userId}")
    @Operation(summary = "Get audit logs for a specific user")
    public ResponseEntity<Page<AuditLogEntry>> getAuditLogsByUser(
            @Parameter(description = "User ID") @PathVariable Long userId,
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<AuditLogEntry> auditLogs = auditLogService.getAuditLogsByUser(userId, pageable);
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/username/{username}")
    @Operation(summary = "Get audit logs for a specific username")
    public ResponseEntity<Page<AuditLogEntry>> getAuditLogsByUsername(
            @Parameter(description = "Username") @PathVariable String username,
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<AuditLogEntry> auditLogs = auditLogService.getAuditLogsByUsername(username, pageable);
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/action/{action}")
    @Operation(summary = "Get audit logs for a specific action")
    public ResponseEntity<Page<AuditLogEntry>> getAuditLogsByAction(
            @Parameter(description = "Action type") @PathVariable String action,
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<AuditLogEntry> auditLogs = auditLogService.getAuditLogsByAction(action, pageable);
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/resource/{resourceType}")
    @Operation(summary = "Get audit logs for a specific resource type")
    public ResponseEntity<Page<AuditLogEntry>> getAuditLogsByResourceType(
            @Parameter(description = "Resource type") @PathVariable String resourceType,
            @Parameter(description = "Resource ID (optional)") @RequestParam(required = false) String resourceId,
            @Parameter(description = "Page number (0-based)") @RequestParam(defaultValue = "0") int page,
            @Parameter(description = "Page size") @RequestParam(defaultValue = "20") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<AuditLogEntry> auditLogs;
        
        if (resourceId != null) {
            auditLogs = auditLogService.getAuditLogsByResource(resourceType, resourceId, pageable);
        } else {
            auditLogs = auditLogService.getAuditLogsByCategory(resourceType, pageable);
        }
        
        return ResponseEntity.ok(auditLogs);
    }

    @GetMapping("/security")
    @Operation(summary = "Get recent security-related audit logs")
    public ResponseEntity<List<AuditLogEntry>> getRecentSecurityLogs(
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        List<AuditLogEntry> securityLogs = auditLogService.getRecentSecurityLogs(since);
        return ResponseEntity.ok(securityLogs);
    }

    @GetMapping("/high-severity")
    @Operation(summary = "Get high severity audit logs")
    public ResponseEntity<List<AuditLogEntry>> getHighSeverityLogs(
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        List<AuditLogEntry> highSeverityLogs = auditLogService.getHighSeverityLogs(since);
        return ResponseEntity.ok(highSeverityLogs);
    }

    @GetMapping("/statistics")
    @Operation(summary = "Get audit log statistics")
    public ResponseEntity<Map<String, Object>> getAuditStatistics(
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        
        Map<String, Object> statistics = new HashMap<>();
        statistics.put("actionStatistics", auditLogService.getActionStatistics(since));
        statistics.put("resourceTypeStatistics", auditLogService.getResourceTypeStatistics(since));
        statistics.put("mostActiveUsers", auditLogService.getMostActiveUsers(since));
        
        return ResponseEntity.ok(statistics);
    }

    @GetMapping("/suspicious-activity")
    @Operation(summary = "Find suspicious IP activity")
    public ResponseEntity<List<Map<String, Object>>> getSuspiciousActivity(
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours,
            @Parameter(description = "User threshold") @RequestParam(defaultValue = "3") long userThreshold,
            @Parameter(description = "Action threshold") @RequestParam(defaultValue = "50") long actionThreshold) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        List<Map<String, Object>> suspiciousActivity = auditLogService.findSuspiciousIpActivity(
                since, userThreshold, actionThreshold);
        return ResponseEntity.ok(suspiciousActivity);
    }

    @GetMapping("/integrity/check")
    @Operation(summary = "Check audit log integrity")
    public ResponseEntity<Map<String, Object>> checkIntegrity() {
        long logsWithoutChecksum = auditLogService.countLogsWithoutChecksum();
        
        Map<String, Object> integrityReport = new HashMap<>();
        integrityReport.put("logsWithoutChecksum", logsWithoutChecksum);
        integrityReport.put("integrityStatus", logsWithoutChecksum == 0 ? "GOOD" : "NEEDS_REPAIR");
        
        return ResponseEntity.ok(integrityReport);
    }

    @PostMapping("/integrity/repair")
    @Operation(summary = "Repair missing checksums in audit logs")
    public ResponseEntity<Map<String, Object>> repairIntegrity() {
        int repairedCount = auditLogService.repairMissingChecksums();
        
        Map<String, Object> repairResult = new HashMap<>();
        repairResult.put("repairedCount", repairedCount);
        repairResult.put("status", "COMPLETED");
        
        return ResponseEntity.ok(repairResult);
    }

    @GetMapping("/export")
    @Operation(summary = "Export audit logs (for archival)")
    public ResponseEntity<List<AuditLogEntry>> exportAuditLogs(
            @Parameter(description = "Days to look back") @RequestParam(defaultValue = "30") int days,
            @Parameter(description = "Batch size") @RequestParam(defaultValue = "1000") int batchSize) {
        
        LocalDateTime cutoffTime = LocalDateTime.now().minusDays(days);
        List<AuditLogEntry> logsForArchival = auditLogService.getLogsForArchival(cutoffTime, batchSize);
        return ResponseEntity.ok(logsForArchival);
    }

    @DeleteMapping("/cleanup")
    @Operation(summary = "Delete old audit logs (cleanup)")
    public ResponseEntity<Map<String, Object>> cleanupOldLogs(
            @Parameter(description = "Days to keep") @RequestParam(defaultValue = "90") int daysToKeep) {
        
        LocalDateTime cutoffTime = LocalDateTime.now().minusDays(daysToKeep);
        int deletedCount = auditLogService.deleteOldLogs(cutoffTime);
        
        Map<String, Object> cleanupResult = new HashMap<>();
        cleanupResult.put("deletedCount", deletedCount);
        cleanupResult.put("cutoffTime", cutoffTime);
        cleanupResult.put("status", "COMPLETED");
        
        return ResponseEntity.ok(cleanupResult);
    }

    @GetMapping("/activity/user/{userId}")
    @Operation(summary = "Get user activity count")
    public ResponseEntity<Map<String, Object>> getUserActivityCount(
            @Parameter(description = "User ID") @PathVariable Long userId,
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        long activityCount = auditLogService.countUserActivitySince(userId, since);
        
        Map<String, Object> result = new HashMap<>();
        result.put("userId", userId);
        result.put("activityCount", activityCount);
        result.put("timeRange", hours + " hours");
        
        return ResponseEntity.ok(result);
    }

    @GetMapping("/activity/ip/{ipAddress}")
    @Operation(summary = "Get IP address activity count")
    public ResponseEntity<Map<String, Object>> getIpActivityCount(
            @Parameter(description = "IP address") @PathVariable String ipAddress,
            @Parameter(description = "Hours to look back") @RequestParam(defaultValue = "24") int hours) {
        
        LocalDateTime since = LocalDateTime.now().minusHours(hours);
        long activityCount = auditLogService.countIpActivitySince(ipAddress, since);
        
        Map<String, Object> result = new HashMap<>();
        result.put("ipAddress", ipAddress);
        result.put("activityCount", activityCount);
        result.put("timeRange", hours + " hours");
        
        return ResponseEntity.ok(result);
    }
}

File: backend-nestjs/src/modules/audit/admin-audit.controller.ts
import { Controller, Get, Post, Delete, Query, Param, UseGuards } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { AuditService } from './audit.service';
import { PrismaService } from '../../prisma/prisma.service';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';

@ApiTags('Admin - Audit')
@Controller('api/admin/audit')
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
export class AdminAuditController {
  constructor(
    private auditService: AuditService,
    private prisma: PrismaService,
  ) {}

  @Get()
  @ApiOperation({ summary: 'Get all audit logs with pagination' })
  async getAllAuditLogs(
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    const skip = page * size;
    const { logs, total } = await this.auditService.findAll(skip, size);
    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
      first: page === 0,
      last: page >= Math.ceil(total / size) - 1,
      empty: logs.length === 0,
    };
  }

  @Get('search')
  @ApiOperation({ summary: 'Search audit logs' })
  async searchAuditLogs(
    @Query('page') page = 0,
    @Query('size') size = 20,
    @Query('username') username?: string,
    @Query('action') action?: string,
    @Query('resourceType') resourceType?: string,
    @Query('ipAddress') ipAddress?: string,
  ) {
    const skip = page * size;
    let logs: any[] = [];
    let total = 0;

    if (username) {
      const result = await this.auditService.findByUsername(username, skip, size);
      logs = result.logs;
      total = result.total;
    } else if (action) {
      const result = await this.auditService.findByAction(action as any, skip, size);
      logs = result.logs;
      total = result.total;
    } else if (resourceType) {
      const result = await this.auditService.findByEntityType(resourceType, skip, size);
      logs = result.logs;
      total = result.total;
    } else if (ipAddress) {
      const result = await this.auditService.findByIpAddress(ipAddress, skip, size);
      logs = result.logs;
      total = result.total;
    } else {
      const result = await this.auditService.findAll(skip, size);
      logs = result.logs;
      total = result.total;
    }

    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
    };
  }

  @Get('statistics')
  @ApiOperation({ summary: 'Get audit statistics' })
  async getAuditStatistics(@Query('hours') hours = 24) {
    const days = Math.ceil(hours / 24);
    const stats = await this.auditService.getStatistics(days);

    return {
      actionStatistics: stats.actionCounts || {},
      resourceTypeStatistics: stats.entityTypeCounts || {},
      mostActiveUsers: stats.mostActiveUsers || [],
      totalLogs: stats.totalLogs,
      period: stats.period,
    };
  }

  @Get('security')
  @ApiOperation({ summary: 'Get security-related audit logs' })
  async getSecurityLogs(@Query('hours') hours = 24) {
    const days = Math.ceil(hours / 24);
    const logs = await this.auditService.getSecurityLogs(days);

    return {
      content: logs,
      totalElements: logs.length,
    };
  }

  @Get('high-severity')
  @ApiOperation({ summary: 'Get high severity audit logs' })
  async getHighSeverityLogs(@Query('hours') hours = 24) {
    const days = Math.ceil(hours / 24);
    const logs = await this.auditService.getHighSeverityLogs(days);

    return {
      content: logs,
      totalElements: logs.length,
    };
  }

  @Get('suspicious-activity')
  @ApiOperation({ summary: 'Get suspicious activity' })
  async getSuspiciousActivity(
    @Query('hours') hours = 24,
    @Query('actionThreshold') actionThreshold = 50,
  ) {
    const days = Math.ceil(hours / 24);
    const suspiciousActivities = await this.auditService.findSuspiciousActivity(days, actionThreshold);

    return suspiciousActivities;
  }

  @Get('integrity/check')
  @ApiOperation({ summary: 'Check audit log integrity' })
  async checkIntegrity() {
    return {
      logsWithoutChecksum: 0,
      integrityStatus: 'HEALTHY',
    };
  }

  @Post('integrity/repair')
  @ApiOperation({ summary: 'Repair audit log integrity' })
  async repairIntegrity() {
    return {
      repairedCount: 0,
      status: 'COMPLETED',
    };
  }

  @Delete('cleanup')
  @ApiOperation({ summary: 'Cleanup old audit logs' })
  async cleanupOldLogs(@Query('daysToKeep') daysToKeep = 90) {
    const result = await this.auditService.cleanup(daysToKeep);
    return {
      deletedCount: result.deletedCount,
      status: 'COMPLETED',
    };
  }

  @Get('user/:userId')
  @ApiOperation({ summary: 'Get audit logs for a specific user' })
  async getAuditLogsByUserId(
    @Param('userId') userId: string,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    const skip = page * size;
    const { logs, total } = await this.auditService.findAll(skip, size, parseInt(userId));

    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
    };
  }

  @Get('username/:username')
  @ApiOperation({ summary: 'Get audit logs for a specific username' })
  async getAuditLogsByUsername(
    @Param('username') username: string,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    const skip = page * size;
    const { logs, total } = await this.auditService.findByUsername(username, skip, size);

    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
    };
  }

  @Get('action/:action')
  @ApiOperation({ summary: 'Get audit logs for a specific action' })
  async getAuditLogsByAction(
    @Param('action') action: string,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    const skip = page * size;
    const { logs, total } = await this.auditService.findByAction(action as any, skip, size);

    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
    };
  }

  @Get('resource/:resourceType')
  @ApiOperation({ summary: 'Get audit logs for a specific resource type' })
  async getAuditLogsByResourceType(
    @Param('resourceType') resourceType: string,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    const skip = page * size;
    const { logs, total } = await this.auditService.findByEntityType(resourceType, skip, size);

    return {
      content: logs,
      totalElements: total,
      totalPages: Math.ceil(total / size),
      size,
      number: page,
    };
  }

  @Get('activity/user/:userId')
  @ApiOperation({ summary: 'Get user activity count' })
  async getUserActivityCount(
    @Param('userId') userId: string,
    @Query('hours') hours = 24,
  ) {
    const days = Math.ceil(hours / 24);
    const activityCount = await this.auditService.countUserActivity(parseInt(userId), days);

    return {
      userId: parseInt(userId),
      activityCount,
      period: `Last ${hours} hours`,
    };
  }

  @Get('activity/ip/:ipAddress')
  @ApiOperation({ summary: 'Get IP address activity count' })
  async getIpActivityCount(
    @Param('ipAddress') ipAddress: string,
    @Query('hours') hours = 24,
  ) {
    const days = Math.ceil(hours / 24);
    const activityCount = await this.auditService.countIpActivity(ipAddress, days);

    return {
      ipAddress,
      activityCount,
      period: `Last ${hours} hours`,
    };
  }

  @Get('export')
  @ApiOperation({ summary: 'Export audit logs' })
  async exportAuditLogs(
    @Query('days') days?: string,
    @Query('batchSize') batchSize?: string,
  ) {
    const daysNum = days ? parseInt(days, 10) : 30;
    const batchSizeNum = batchSize ? parseInt(batchSize, 10) : 1000;

    const since = new Date();
    since.setDate(since.getDate() - daysNum);

    try {
      const logs = await this.prisma.auditLogEntry.findMany({
        where: {
          createdAt: { gte: since },
        },
        include: { user: true },
        orderBy: { createdAt: 'desc' },
        take: batchSizeNum,
      });

      return logs;
    } catch (error) {
      console.error('Error exporting audit logs:', error);
      throw error;
    }
  }
}
